packages/shopify-app-remix/src/server/authenticate/admin/embedded-authenticate.ts

```diff
@@ -0,0 +1,245 @@
+    const sessionTokenParam = url.searchParams.get(SESSION_TOKEN_PARAM)!;
+    const sessionTokenString = sessionTokenHeader || sessionTokenParam;
+
+    logger.info('Authenticating admin request');
+
+    if (sessionTokenParam) {
+      await validateUrlParams(request, {api, logger, config});
+      await ensureAppIsEmbeddedIfRequired(request, {api, logger, config});
+    }
+
+    if (sessionTokenString) {
+      try {
+        const sessionToken = await validateSessionTokenUncaught(
+          {api, logger, config},
+          sessionTokenString,
+        );
+
+        return await this.getAccessToken(
+          request,
+          sessionTokenString,
+          sessionToken,
+        );
+      } catch (error) {
+        if (
+          error instanceof InvalidJwtError ||
+          (error instanceof HttpResponseError &&
+            error.response.code === 400 &&
+            error.response.body?.error === 'invalid_subject_token')
+        ) {
+          console.log(`CAUGHT ${JSON.stringify(error)}`);
+          throw this.handleInvalidSession(request);
+        }
+        throw error;
+      }
+    } else {
+      logger.debug(
+        'Missing session token in search params, going to bounce page',
+      );
+      throw this.handleInvalidSession(request);
+    }
+  }
+
+  private async getAccessToken(
+    request: Request,
+    sessionToken: string,
+    payload: JwtPayload,
+  ) {
+    const {config, logger, api} = this;
+
+    const sessionId = await api.session.getCurrentId({
+      isOnline: config.useOnlineTokens,
+      rawRequest: request,
+    });
+
+    const dest = new URL(payload.dest);
+    const shop = dest.hostname;
+
+    if (sessionId) {
+      logger.debug(`SESSION ID: ${sessionId}`);
+      const persistedSession = await config.sessionStorage.loadSession(
+        sessionId,
+      );
+
+      if (persistedSession) {
+        logger.debug(`Reusing existing token: ${persistedSession.accessToken}`);
+
+        if (persistedSession.isScopeChanged(config.scopes)) {
+          config.sessionStorage.deleteSession(persistedSession.id);
+
+          this.redirectToInstall(request, shop);
+        }
+
+        if (!persistedSession.isExpired()) {
+          return {session: persistedSession};
+        }
+      }
+    }
+
+    logger.debug('Requesting token exchange');
+
+    const {session} = await api.auth.tokenExchange({
+      sessionToken,
+      shop,
+      isOnline: config.useOnlineTokens,
+    });
+
+    logger.debug(`RECEIVED TOKEN: ${session.accessToken}`);
+
+    await config.sessionStorage.storeSession(session);
+
+    return {session, token: payload};
+  }
+
+  private handleInvalidSession(request: Request) {
+    const {api, logger, config} = this;
+
+    const isXhrRequest = request.headers.get('authorization');
+    if (!isXhrRequest) {
+      return redirectToBouncePage(new URL(request.url), {api, logger, config});
+    }
+
+    return new Response(undefined, {
```

Comment:
For an XHR request made via App Bridge's augmented fetch, return a `401` with the `X-Shopify-Invalid-Session` header.